### PR TITLE
New UI tech test with vite

### PR DESCRIPTION
**Pull request by kml1990**
# New UI tech test. 

## App info
The app is using vite to bootstrap react and server using `npm run dev` command.

App is written in `TypeScript` but config is set to `strict: false` which allows to write `JS` code without a need to use `TS`

I added test as a test runner for the app with one simple test (similar to what we had).

## UI
App UI is very basic atm. Just a header with input to get them started.

Screenshot 2023-11-20 at 16 58 02

## Code
Very basic starter with two components created to get them started. Fetch has been removed to see how they are handling it themselves (Need to observe if it's still feasible to complete part1 of the test without fetch not being there) 

Screenshot 2023-11-20 at 17 21 06

## Endpoints
We now have two endpoints `/api/tags` and `/api/cars/` which will allow candidates to create a similar experience to Pinterest. 
Both endpoints allow a `tag` query string to be passed.

```
/* 
 * Available endpoints
 * http://localhost:8000/api/tags - to return all tags in
 * http://localhost:8000/api/tags?tag=fe - to return matching tags
 * http://localhost:8000/api/cars - to return all cars
 * http://localhost:8000/api/cars?tag=ferrari - to return matching cars
 */
```

Adding `tag` query to `/api/tags` endpoint will match all available tags using `startsWith` function so adding `/api/tags?tag=f` will return `[ferrari, ford, forest]` but `/api/tags?tag=fe` will only return `[ferrari]`. Have a look on the video attached for demo.

https://user-images.githubusercontent.com/9830035/224315617-111fcb42-0367-4cb8-a7ce-16429f09b3ec.mov

Adding `tag` query to `api/cars` endpoint will match all the cars that include selected tag from `tags` endpoint, so `/api/cars?tag=ford` will return all fords available. Have a look on the video attached for demo.

https://user-images.githubusercontent.com/9830035/224315634-9a371f8a-b2e7-4289-8036-03e19e8a8186.mov

## Readme
I have rewritten readme slightly, let me know if it is making sense 👍

### CODE REVIEW

**Review by dhadfieldmotorway — DISMISSED** [review]
Small nit but non blocking

File: tsconfig.json [hunk 1]
```diff
@@ -0,0 +1,20 @@
...
+    "esModuleInterop": false,
+    "allowSyntheticDefaultImports": true,
+    "strict": false,
+    "forceConsistentCasingInFileNames": true,
+    "module": "ESNext",
+    "moduleResolution": "Node",
+    "resolveJsonModule": true,
+    "isolatedModules": true,
+    "noEmit": true,
+    "jsx": "react-jsx",
```
[the remark this answers is not in the document]
Comment (kml1990) [reply tsconfig.json]: Good shout 👍

**Review by mattgordomw — COMMENTED** [review]
Amazing work @kml1990! I have one question which I have added. Will get it running locally as well to check it out

**Comment by szympajka** [comment]
I'll have a look on Monday

**Review by dhadfieldmotorway — APPROVED** [review]

**Review by dalyr95 — APPROVED** [review]
test-your-strength

Works well and the ReadMe makes sense. Nice one! 🌮

**Review by szympajka — APPROVED** [review]

File: README.md [hunk 2]
```diff
@@ -7,72 +7,172 @@ Welcome to the Motorway UI technical test. This test focuses on user experience,
...
+There are two stages to this task depending on the type of challenge you're taking.
+
+#### Live challenge (1 hour)
+[Task one](#1-realtime-search).
+
+#### Take home challenge (2 hours)
+Task one and attempt [task two](#2-realtime-search-results). We appreciate that your time is valuable and recommend you not spend more than 2 hours on these tasks. Focusing on one small thing well or a bigger scope as a proof of concept, both approaches are valid. 
+
+### 1. Realtime search
+We have an tags api endpoint at `http://localhost:8000/tags` which returns an array of all tags available in the format:
```
[the remark this answers is not in the document]
Comment (szympajka) [reply README.md]: fixed

File: README.md [hunk 3]
```diff
@@ -7,72 +7,172 @@ Welcome to the Motorway UI technical test. This test focuses on user experience,
...
+
+#### Outcomes
+To have a type ahead search functionality like [Pinterest](https://www.pinterest.co.uk/ideas/) which can query the above API and return selectable results. 
+- Display tags available in a form of dropdown
+- Allow available tags to be selected
+- Performant fast results
+- The dataset is static but **assuming** it could be dynamic would be beneficial
+- **Not exposing the whole dataset** would be beneficial 
+
+```
```
[the remark this answers is not in the document]
Comment (szympajka) [reply README.md]: fixed

**Review by mattgordomw — APPROVED** [review]
Nice!

File: src/server/server.ts [hunk 4]
```diff
@@ -23,6 +23,11 @@ app.get('/api/cars', ({ query }: Request<{}, {}, {}, Query>, res) => {
   const { tag } = query;
   const carsByTag = getCarsByTag(tag, cars);
 
+  if (carsByTag.length === 0) {
+    res.json(["No Results"]);
```
Comment (dalyr95) [inline src/server/server.ts:27]: what's the existing response before this @kml1990 ? I think an empty array would be the norm

Comment (kml1990) [reply src/server/server.ts:27]: Agree @dalyr95 👍  Empty array would definitely be better here 👍 Will change that 👍

**Comment by kml1990** [comment]
@mattgordomw @szympajka @dalyr95 @dhadfieldmotorway I have modified this PR slightly. Changes were made based on the feedback received from @szympajka. 

We now have figma designs available for candidates. Attached link to [figma](https://www.figma.com/file/AQ6dwlbpusgQzBpJyAqcVq/Recruitment?type=design&node-id=37-932&mode=design) in readme. Designs are password protected (Motorway).

The initial setup is not responsive but we have included the breakpoints that can be used to make it responsive.

Please review the code, try to run this locally and let me know your thoughts 👍  Thanks.

**Review by szympajka — APPROVED** [review]
Great work @kml1990!

I played with the test a bit, and feels like we refined it really well! :) Good work! 🙇

**Review by mattgordomw — APPROVED** [review]
Top work @kml1990 Looking forward to seeing these live and also take home versions!

**Comment by kml1990** [comment]
@szympajka @mattgordomw fixed the snapshot. Its ready for re-review 👍

File: src/components/app/__snapshots__/App.test.tsx.snap [hunk 5]
```diff
@@ -0,0 +1,474 @@
...
+  "queryAllByTestId": [Function],
+  "queryAllByText": [Function],
+  "queryAllByTitle": [Function],
+  "queryByAltText": [Function],
+  "queryByDisplayValue": [Function],
+  "queryByLabelText": [Function],
+  "queryByPlaceholderText": [Function],
+  "queryByRole": [Function],
+  "queryByTestId": [Function],
+  "queryByText": [Function],
```
Comment (szympajka) [inline src/components/app/__snapshots__/App.test.tsx.snap:469]: hey @kml1990, we are still capturing all those mock fns which arent usually what you want

**Comment by kml1990** [comment]
Should be fixed now @szympajka.I realised that there were two snapshots for app component which remained after I moved app component to its own folder 🤯

**Review by szympajka — APPROVED** [review]

**Comment by szympajka** [comment]
What do we do with this PR now @kml1990?

**Pull request closed without merging**